### ISSUE TITLE
Status update on Incremental and Grid Search

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask-ml#206]
Quick status update: I wanted to explore a workflow that would use scikit-learn
as much as possible. We'd use scikit-learn for all the hyper-parameter
optimization and the actual training. Dask would just provide the large
arrays.

The end goal is something as close as possible to `GridSearchCV(SGDClassifier())` trained on a larger-than-memory dataset.

```python
X, y = load_dask_arrays()

clf = sklearn.linear_model.SGDClassifier()
gs = sklearn.model_selection.GridSearchCV(clf, param_grid)
gs.fit(X, y)
```

First, we have to avoid passing a large Dask array to `SGDClassifier.fit`, as it
would be converted to an ndarray. So we wrap it with `Incremental`, which passes
blocks to the `SGDClassifier.partial_fit`:

```python
import dask_ml.wrappers

X, y = load_dask_arrays()

clf = sklearn.linear_model.SGDClassifier()
inc = dsak_ml.wrappers.Incremental(clf)
gs = sklearn.model_selection.GridSearchCV(clf, param_grid)
gs.fit(X, y)
```

At this point, `inc.fit(X, y)` works fine (I'm currently debugging unexpectedly
high memory usage, but let's ignore that for now), but there's a subtle issue with
`inc.score(X, y)`, which surfaces in `GridSeachCV`. By default, a pass-through
scorer is used, which is `SGDClassifier.score`. This ends up using
`sklearn.metrics.accuracy_score`, which end up converting the test Dask arrays
into large ndarrays on a single worker. This is captured in #200.

A workaround to the scoring issue is to manually pass a scorer that is able to
work well with Dask arrays. We've implemented a few in Dask-ML:

```python
import dask_ml.wrappers
import dask_ml.metrics
from sklearn.metrics import make_scorer

X, y = load_dask_arrays()
scorer = make_scorer(dask_ml.metrics.accuracy_score)

clf  = sklearn.linear_model.SGDClassifier()
inc = dask_ml.wrappers.Incremental(clf, scoring=scorer)
gs = sklearn.model_selection.GridSearchCV(clf, param_grid)
gs.fit(X, y)
```

This gets us serial hyper-parameter optimization on larger-than-memory Dask
arrays. To do things in parallel, we can use the distributed backend.

```python
import dask_ml.joblib
from sklearn.externals import joblib


with joblib.parallel_backend("dask"):
    gs.fit(X, y)
```

I'll post benchmarks later when I've run them.

**Comment by TomAugspurger** [comment]
cc @stsievert

**Comment by TomAugspurger** [comment]
Here's a gist showing out of core SGD https://gist.github.com/245d492f95f2f4d6519dffdc3bd72ed9 (no hyperparameter optimization yet).

Here's the scheduler UI on a 24 GB dataset (my laptop has 16GB of RAM): https://streamable.com/1goff

We go through all of the zarr reading (including some re-writing to disk since the dataset is too large), before moving onto the first fit.

Ideally, we would start fitting the first block, and releasing it from memory, as soon as it's available. I haven't dug into why we don't do this yet.

**Comment by TomAugspurger** [comment]
This is the graph for a smaller problem.

![mydask](https://user-images.githubusercontent.com/1312546/41178268-64b06a68-6b2c-11e8-99cc-7a9c60a14465.png)

generated on

```
make_data(n_samples=50_000, n_features=500, chunks=5_000, random_state=0)
from dask_ml._partial import fit

val = fit(clf, X, y, classes=[0, 1], compute=False)
val.visualize(rankdir='LR', color='order');
```

**Comment by mrocklin** [comment]
Interesting.  It's preferring the outer nodes first instead of the inner ones.  It would be good to investigate why this is.

**Comment by mrocklin** [comment]
If you want to dive into dask.order that is

**Comment by TomAugspurger** [comment]
Yes, I would like to debug this, though I may ping you if I get stuck :)

**Comment by mrocklin** [comment]
The tests for ordering are probably a good place to start.  The test
docstrings might be informative about the kinds of cases we consider.

**Comment by TomAugspurger** [comment]
Interestingly, we take the right path for a totally random dataset from `da.random.normal`

```python
from dask_ml._partial import fit

X = da.random.uniform(size=(50_000, 500), chunks=5_000)
y = da.random.uniform(size=(50_000,), chunks=5_000)
val = fit(clf, X, y, classes=[0, 1], compute=False)

val.visualize('random-uniform', rankdir='LR', color='order');
```

![random-uniform](https://user-images.githubusercontent.com/1312546/41179062-0062e3d0-6b2f-11e8-95e4-0d7a0d81b7ac.png)

**Comment by TomAugspurger** [comment]
Some notes comparing those two, `da.random.uniform` (good) and zarr (bad).

Identical up on the initial pass through `stack` up until the items
`(da.random.uniform, 0, 0)` / `(X.zarr, 0, 0)`. They have different dependents.

- random uniform has no deps
- `(X.zarr, 0, 0)` has one: `{'zarr-X.zarr'}`

So we diverge here. The `da.random.uniform` version correctly goes `(X, 0, 0), (y, 0), fit-0, ...` at this point.
The zarr version throws `zarr-X.zarr` on the stack and then starts working through `(X, 9, 0), (X, 8, 0), ... (y, 9), (y, 8), ... fit-0, fit-1, ..., fit-9)`

**Comment by TomAugspurger** [comment]
Here's the some trimmed total_dependents:

```
# random version
{('da.random.uniform-1fbca3c0347dfde415f84ec4a950bd61', 0, 0): 11,
# ...
 ('da.random.uniform-1fbca3c0347dfde415f84ec4a950bd61', 9, 0): 2,
 ('da.random.uniform-fce7cf97942af8ebd4353a56e37d6783', 0): 11,
# ...
 ('da.random.uniform-fce7cf97942af8ebd4353a56e37d6783', 9): 2,
 ('fit-630d7694014736f7c87d2604689f65d9', -1): 11,
 ('fit-630d7694014736f7c87d2604689f65d9', 0): 10,
# ...
 ('fit-630d7694014736f7c87d2604689f65d9', 9): 1}
```

```
# Zarr version
{'zarr-X.zarr': 66,
 'zarr-y.zarr': 66,
 ('fit-e1fcb35a27c9528708d6eb0ab04b0f47', -1): 11,
 ('fit-e1fcb35a27c9528708d6eb0ab04b0f47', 0): 10,
...
 ('fit-e1fcb35a27c9528708d6eb0ab04b0f47', 9): 1,
 ('zarr-X.zarr', 0, 0): 11,
...
 ('zarr-X.zarr', 9, 0): 2,
 ('zarr-y.zarr', 0): 11,
...
 ('zarr-y.zarr', 9): 2}
```

I wonder if it's as simple as `Zarr-X.zarr` has lots of dependents, so lets go through that first...

**Comment by mrocklin** [comment]
I suspect that in practice data would be coming from a single source and then split apart rather than two separate sources.  Task ordering might behave better in this situation.

**Comment by jakirkham** [comment]
> Interesting. It's preferring the outer nodes first instead of the inner ones. It would be good to investigate why this is.

This is looking reminiscent of the issues we have seen with ordering in `map_overlap`. ( https://github.com/dask/dask/issues/3554 ) Is it possible they share a common cause? Don't know too much about how ordering works currently. So apologize if this is a naive question.

**Comment by TomAugspurger** [comment]
> I suspect that in practice data would be coming from a single source and then split apart rather than two separate sources. Task ordering might behave better in this situation.

Not for this simple example unfortunately (X and y in the same Zarr file, read in together, and then split).

![xy-zarr](https://user-images.githubusercontent.com/1312546/41182792-d001886e-6b3c-11e8-9057-06dba697ef6f.png)


```python
def make_data(n_samples=1_000_000, n_features=1_000, chunks=5_000, random_state=None):
    X, y = make_classification(n_samples=n_samples, n_features=n_features, chunks=chunks,
                               random_state=random_state)
    y = y.reshape(-1, 1)
    Xy = da.concatenate([X, y], axis=1)
    print("Generating random X-y pair")
    print(f"{X.nbytes/1e9} GB")
    print(f"{X.numblocks} blocks")
    da.to_zarr(Xy, "Xy.zarr", overwrite=True)

def load_data():
    Xy = da.from_zarr("Xy.zarr")
    y = Xy[:, -1]
    X = Xy[:, :-1]
    return X, y

X, y = load_data()
val = fit(clf, X, y, classes=[0, 1], compute=False)
val.visualize('xy-zarr', rankdir='LR', color='order');
```

**Comment by mrocklin** [comment]
That seems like an excellent fail case.  My next step here would be to make an minimal abstract test, similar to what is found in `dask/tests/test_order.py` and then tweak `order.py` to see what is going on.

**Comment by TomAugspurger** [comment]
Ok, I think this is a fairly minimal example:

```python
def f(*args):
    pass


dsk1 = {
    'X0': 0,
    'y0': 0,
    'b0': 0,
    'b1': (f, 'b0', 'X0', 'y0'),
    'X1': 1,
    'y1': 1,
    'b2': (f, 'b1', 'X1', 'y1')
}

dsk2 = {
    'Xy': 0,
    'X0': (f, 'Xy', 0, 0),
    'y0': (f, 'Xy', 0, 1),
    'b0': 0,
    'b1': (f, 'b0', 'X0', 'y0'),
    'X1': (f, 'Xy', 1, 0),
    'y1': (f, 'Xy', 1, 1),
    'b2': (f, 'b1', 'X1', 'y1')
}
```

**dsk1** This is handled well. The blocks of `X` don't have a common ancestor. This is the "shallow" case, and we make the right choice.


![mydask](https://user-images.githubusercontent.com/1312546/41206572-9ac9c662-6ccb-11e8-919f-7f244e3cff24.png)


**dsk2** is not handled well. The blocks of `X` (and `y`) share a common ancestor. We choose to go down the outer edge first 

![mydask](https://user-images.githubusercontent.com/1312546/41206573-a1f4e836-6ccb-11e8-9eb3-e4c2a008e4d0.png)

I'll start playing with the ordering heuristics.

**Comment by TomAugspurger** [comment]
> make an minimal abstract test

(That was a useful exercise)

**Comment by mrocklin** [comment]
Have fun!  This will undoubtedly introduce you to the many other graph motifs like this that we try to handle optimally.  My experience is that in trying to touch any part of this problem you undoubtedly end up having to load the entire thing into your brain.  This isn't necessarily bad, it's an interesting and reasonably sized problem.

**Comment by TomAugspurger** [comment]
https://github.com/dask/dask/pull/3588 solves the ordering issue.
